{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7f9bada",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Display all the columns of dataframes\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e1ad75",
   "metadata": {},
   "source": [
    "# Pick an RBP to use for a toy example\n",
    "To do this, look at Supp. Data 4 in [Van Nostrand et al. (2020)](https://doi.org/10.1038/s41586-020-2077-3). \n",
    "Pick an RBP that...\n",
    "1. Has many total IDR peaks\n",
    "2. Has many peaks near 5' and 3' splice sites\n",
    "3. If possible also has peaks in distal intron?\n",
    "\n",
    "Choose KHSRP (K562 dataset has more peaks than HepG2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06bb1b7c",
   "metadata": {},
   "source": [
    "Open pyranges from `str_near_splice.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c76620a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13623</td>\n",
       "      <td>13635</td>\n",
       "      <td>ACAGGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CAGGCA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1749</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>uc010nxr.1</td>\n",
       "      <td>uc010nxr.1</td>\n",
       "      <td>3</td>\n",
       "      <td>uc010nxr.1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>322376</td>\n",
       "      <td>322388</td>\n",
       "      <td>AAGAAT</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GAATAA</td>\n",
       "      <td>uc009vjk.2</td>\n",
       "      <td>322037</td>\n",
       "      <td>326938</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>339</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>322036</td>\n",
       "      <td>322228</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>C9J4L2</td>\n",
       "      <td>uc009vjk.2</td>\n",
       "      <td>1</td>\n",
       "      <td>uc009vjk.2.1</td>\n",
       "      <td>C9J4L2</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>877247</td>\n",
       "      <td>877259</td>\n",
       "      <td>AGGGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGGG</td>\n",
       "      <td>uc001abx.1</td>\n",
       "      <td>871152</td>\n",
       "      <td>879961</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>6095</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>877515</td>\n",
       "      <td>877631</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>H7BY14</td>\n",
       "      <td>uc031pjy.1</td>\n",
       "      <td>8</td>\n",
       "      <td>uc031pjy.1.8</td>\n",
       "      <td>H7BY14</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879886</td>\n",
       "      <td>879906</td>\n",
       "      <td>AGCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>CCAGC</td>\n",
       "      <td>uc001abx.1</td>\n",
       "      <td>871152</td>\n",
       "      <td>879961</td>\n",
       "      <td>+</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>8734</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>879287</td>\n",
       "      <td>879961</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Q96NU1</td>\n",
       "      <td>uc031pjt.1</td>\n",
       "      <td>12</td>\n",
       "      <td>uc031pjt.1.12</td>\n",
       "      <td>Q96NU1</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>882085</td>\n",
       "      <td>882097</td>\n",
       "      <td>AGAGGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGCAGA</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11821</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>881781</td>\n",
       "      <td>882440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>uc031pkq.1</td>\n",
       "      <td>5</td>\n",
       "      <td>uc031pkq.1.5</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>882362</td>\n",
       "      <td>882374</td>\n",
       "      <td>AGATGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGAGAT</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11544</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>881781</td>\n",
       "      <td>882440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>uc031pkq.1</td>\n",
       "      <td>5</td>\n",
       "      <td>uc031pkq.1.5</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>882442</td>\n",
       "      <td>882456</td>\n",
       "      <td>ACT</td>\n",
       "      <td>14</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>TAC</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11462</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>881781</td>\n",
       "      <td>882440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>uc031pkq.1</td>\n",
       "      <td>5</td>\n",
       "      <td>uc031pkq.1.5</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>888115</td>\n",
       "      <td>888127</td>\n",
       "      <td>ACCCCC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCCA</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>5791</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>887791</td>\n",
       "      <td>887980</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NP_056473</td>\n",
       "      <td>uc001abz.4</td>\n",
       "      <td>10</td>\n",
       "      <td>uc001abz.4.10</td>\n",
       "      <td>NP_056473</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    length | +23   |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+-------|\n",
       "| chr1         |     13623 |     13635 | ACAGGC     |        12 | ...   |\n",
       "| chr1         |    322376 |    322388 | AAGAAT     |        12 | ...   |\n",
       "| chr1         |    877247 |    877259 | AGGGG      |        12 | ...   |\n",
       "| chr1         |    879886 |    879906 | AGCCC      |        20 | ...   |\n",
       "| chr1         |    882085 |    882097 | AGAGGC     |        12 | ...   |\n",
       "| chr1         |    882362 |    882374 | AGATGG     |        12 | ...   |\n",
       "| chr1         |    882442 |    882456 | ACT        |        14 | ...   |\n",
       "| chr1         |    888115 |    888127 | ACCCCC     |        12 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 28 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "23 hidden columns: Strand, num_units, actual_repeat, gene_id, gene_start, gene_stop, ... (+ 17 more.)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "within_300_pr = pr.PyRanges(pd.read_csv(\"within_300_ss_pyranges.csv\"))\n",
    "within_300_pr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e854b7c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>22811</td>\n",
       "      <td>22823</td>\n",
       "      <td>AAAGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGAA</td>\n",
       "      <td>uc001aac.4</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>6547</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>24737</td>\n",
       "      <td>24891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>9</td>\n",
       "      <td>uc009vir.3.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23874</td>\n",
       "      <td>23886</td>\n",
       "      <td>ACCTCC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCTCCA</td>\n",
       "      <td>uc009viq.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>5484</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>24737</td>\n",
       "      <td>24891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>9</td>\n",
       "      <td>uc009vir.3.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>669727</td>\n",
       "      <td>669739</td>\n",
       "      <td>AAC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>AAC</td>\n",
       "      <td>uc009vjm.3</td>\n",
       "      <td>661139</td>\n",
       "      <td>670994</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1255</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>668686</td>\n",
       "      <td>668744</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>B4DGA0</td>\n",
       "      <td>uc009vjm.3</td>\n",
       "      <td>2</td>\n",
       "      <td>uc009vjm.3.2</td>\n",
       "      <td>B4DGA0</td>\n",
       "      <td>984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>676490</td>\n",
       "      <td>676503</td>\n",
       "      <td>AAAAG</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AAAAG</td>\n",
       "      <td>uc001abm.2</td>\n",
       "      <td>674240</td>\n",
       "      <td>679736</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>3233</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>675508</td>\n",
       "      <td>675566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abm.2</td>\n",
       "      <td>uc001abm.2</td>\n",
       "      <td>3</td>\n",
       "      <td>uc001abm.2.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>700958</td>\n",
       "      <td>700970</td>\n",
       "      <td>AAATAC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>TACAAA</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>13098</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>700244</td>\n",
       "      <td>700627</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>1</td>\n",
       "      <td>uc001abo.3.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>701123</td>\n",
       "      <td>701136</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>A</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>12932</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>700244</td>\n",
       "      <td>700627</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>1</td>\n",
       "      <td>uc001abo.3.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>701240</td>\n",
       "      <td>701253</td>\n",
       "      <td>ACTCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCACT</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>12815</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>701708</td>\n",
       "      <td>701767</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>2</td>\n",
       "      <td>uc001abo.3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>702159</td>\n",
       "      <td>702173</td>\n",
       "      <td>AATC</td>\n",
       "      <td>14</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>ATCA</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11895</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>701708</td>\n",
       "      <td>701767</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>2</td>\n",
       "      <td>uc001abo.3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    length | +23   |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+-------|\n",
       "| chr1         |     22811 |     22823 | AAAGG      |        12 | ...   |\n",
       "| chr1         |     23874 |     23886 | ACCTCC     |        12 | ...   |\n",
       "| chr1         |    669727 |    669739 | AAC        |        12 | ...   |\n",
       "| chr1         |    676490 |    676503 | AAAAG      |        13 | ...   |\n",
       "| chr1         |    700958 |    700970 | AAATAC     |        12 | ...   |\n",
       "| chr1         |    701123 |    701136 | A          |        13 | ...   |\n",
       "| chr1         |    701240 |    701253 | ACTCC      |        13 | ...   |\n",
       "| chr1         |    702159 |    702173 | AATC       |        14 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 28 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "23 hidden columns: Strand, num_units, actual_repeat, gene_id, gene_start, gene_stop, ... (+ 17 more.)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_internal_pr = pr.PyRanges(pd.read_csv(\"intron_internal_pyranges.csv\"))\n",
    "intron_internal_pr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887714be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
