{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d56d0027",
   "metadata": {},
   "source": [
    "# Identify STRs that are at splice sites\n",
    "Like the ENCODE paper, define a splice site as being within 100bp of an exon. All of the following uses the GRCh37 alignments (GENCODE v19 annotations), also like the paper. \n",
    "Links to [GENCODE data](https://www.gencodegenes.org/human/release_37lift37.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3107f7fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pyliftover import LiftOver\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "tqdm.pandas(desc = \"Progress\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5889ff22",
   "metadata": {},
   "source": [
    "Load the STR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "29e86a28",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/david/Library/Python/3.7/lib/python/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (9,10,14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10108</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>18</td>\n",
       "      <td>TAACCC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>10149</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10147</td>\n",
       "      <td>10179</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>32</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10172</td>\n",
       "      <td>10184</td>\n",
       "      <td>AACCT</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10177</td>\n",
       "      <td>10233</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>56</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTAAC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End   class  length Strand  num_units actual_repeat  \\\n",
       "0       chr1  10000  10108  AACCCT     108      +         18        TAACCC   \n",
       "1       chr1  10108  10149  AACCCT      41      +          6        AACCCT   \n",
       "2       chr1  10147  10179  AACCCT      32      +          5        CCCTAA   \n",
       "3       chr1  10172  10184   AACCT      12      +          2         CCTAA   \n",
       "4       chr1  10177  10233  AACCCT      56      +          9        CCTAAC   \n",
       "\n",
       "         gene gene_start gene_stop gene_strand  annotation      promoter  \\\n",
       "0  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "1  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "2  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "3  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "4  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "\n",
       "  dist_to_tss  \n",
       "0       -1766  \n",
       "1       -1725  \n",
       "2       -1695  \n",
       "3       -1690  \n",
       "4       -1641  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = [\"Chromosome\", \"Start\", \"End\", \"class\", \n",
    "            \"length\", \"Strand\", \"num_units\", \n",
    "            \"actual_repeat\", \"gene\", \"gene_start\",\n",
    "            \"gene_stop\", \"gene_strand\", \"annotation\",\n",
    "            \"promoter\", \"dist_to_tss\"]\n",
    "dtype_dict = {\"Chromosome\": np.str, \"Start\":np.int, \"End\":np.int, \"length\": np.int, \n",
    "             \"Strand\": np.str} #, \"gene_start\": np.int, \"gene_stop\": np.int, \"dist_to_tss\": np.int}\n",
    "str_df = pd.read_csv(\"data/msdb_data.tsv\", sep = '\\t', dtype = dtype_dict, names = col_names, index_col = False, na_values = \"\")\n",
    "str_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d7d5b3d",
   "metadata": {},
   "source": [
    "Convert MSMB data to GRCh37/hg19 (format of eCLIP data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "07ff1917",
   "metadata": {},
   "outputs": [],
   "source": [
    "lo = LiftOver(\"hg38\", \"hg19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ac00b686",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e4d5db7fa4364aa19d63e6c500e677f9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Progress:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  c\n",
       "0 -4  1\n",
       "1 -3  2"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = pd.DataFrame({\"a\": [1,2], \"b\":[3,4], \"c\": [5,6]})\n",
    "a.head()\n",
    "\n",
    "def testf(x):\n",
    "    return pd.Series({\"a\": x[\"a\"]-5, \"c\": x[\"b\"]-2})\n",
    "\n",
    "a.progress_apply(testf, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4bc5e621",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0c6f4f37d8b24b42ae130e6dc2c135bf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "1st loop:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f623dc57b2044033a39810e16c0bd52e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "02a217d895b24b5299c19e0b8ef4c447",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cecf08516aff42329a1937ac9716604c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from tqdm.notebook import trange, tqdm\n",
    "from time import sleep\n",
    "\n",
    "for i in trange(3, desc='1st loop'):\n",
    "    for j in tqdm(range(100), desc='2nd loop'):\n",
    "        sleep(0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "33f0e201",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "75fc7ddf22024eea8fd8c6fc9039ef5a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Progress:   0%|          | 0/4396272 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__setattr__\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m   5476\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5477\u001b[0;31m             \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5478\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mname\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m         \"\"\"\n\u001b[0;32m--> 536\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_name\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    537\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5460\u001b[0m         ):\n\u001b[0;32m-> 5461\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5462\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute '_name'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-4f9c63b75a99>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mnew_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprogress_apply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconvert_coords\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0mnew_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/tqdm/std.py\u001b[0m in \u001b[0;36minner\u001b[0;34m(df, func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    808\u001b[0m                 \u001b[0;31m# on the df using our wrapper (which provides bar updating)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    809\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 810\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_function\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    811\u001b[0m                 \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    812\u001b[0m                     \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwds)\u001b[0m\n\u001b[1;32m   7766\u001b[0m             \u001b[0mkwds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7767\u001b[0m         )\n\u001b[0;32m-> 7768\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7769\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7770\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapplymap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mOptional\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mget_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    183\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 185\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    186\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    187\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_empty_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    274\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 276\u001b[0;31m         \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    277\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    278\u001b[0m         \u001b[0;31m# wrap results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    288\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m                 \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 290\u001b[0;31m                 \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m                     \u001b[0;31m# If we have a view on v, we need to make a copy because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/tqdm/std.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    803\u001b[0m                     \u001b[0;31m# take a fast or slow code path; so stop when t.total==t.n\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    804\u001b[0m                     \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtotal\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtotal\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 805\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    806\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    807\u001b[0m                 \u001b[0;31m# Apply the provided function (in **kwargs)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-4f9c63b75a99>\u001b[0m in \u001b[0;36mconvert_coords\u001b[0;34m(df_row)\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mdf_row\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Chromosome\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_row\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Strand\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, dtype, name, copy, fastpath)\u001b[0m\n\u001b[1;32m    316\u001b[0m                 \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    317\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mis_dict_like\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 318\u001b[0;31m                 \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_init_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    319\u001b[0m                 \u001b[0mdtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    320\u001b[0m                 \u001b[0mcopy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_init_dict\u001b[0;34m(self, data, index, dtype)\u001b[0m\n\u001b[1;32m    410\u001b[0m         \u001b[0;31m# TODO: passing np.float64 to not break anything yet. See GH-17261\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    411\u001b[0m         s = create_series_with_explicit_dtype(\n\u001b[0;32m--> 412\u001b[0;31m             \u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype_if_empty\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfloat64\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    413\u001b[0m         )\n\u001b[1;32m    414\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/construction.py\u001b[0m in \u001b[0;36mcreate_series_with_explicit_dtype\u001b[0;34m(data, index, dtype, name, copy, fastpath, dtype_if_empty)\u001b[0m\n\u001b[1;32m    665\u001b[0m         \u001b[0mdtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdtype_if_empty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    666\u001b[0m     return Series(\n\u001b[0;32m--> 667\u001b[0;31m         \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfastpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfastpath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    668\u001b[0m     )\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, dtype, name, copy, fastpath)\u001b[0m\n\u001b[1;32m    367\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    368\u001b[0m         \u001b[0mgeneric\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mNDFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 369\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    370\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfastpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.7/lib/python/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__setattr__\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m   5475\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5476\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5477\u001b[0;31m             \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5478\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5479\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def convert_coords(df_row):\n",
    "    new_start = lo.convert_coordinate(df_row['Chromosome'], df_row['Start'], df_row['Strand'])\n",
    "    new_end = lo.convert_coordinate(df_row['Chromosome'], df_row['End'], df_row['Strand'])\n",
    "    cols = [\"Chromosome\", \"Start\", \"End\", \"Strand\"]\n",
    "    if len(new_start) > 0 and len(new_end)>0:\n",
    "        data = [new_start[0][0], new_start[0][1], new_end[0][1], new_start[0][2]]\n",
    "    else:\n",
    "        data = [df_row[\"Chromosome\"], None, None, df_row[\"Strand\"]]\n",
    "    return pd.Series(dict(zip(cols, data)))\n",
    "\n",
    "\n",
    "new_df = str_df.progress_apply(convert_coords, axis = 1)\n",
    "new_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ee7e873a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('chr1', 939880, '+', 20849626768)]\n",
      "[('chr1', 939880, '-', 20849626768)]\n"
     ]
    }
   ],
   "source": [
    "print(lo.convert_coordinate(\"chr1\", 1004500, '+'))\n",
    "print(lo.convert_coordinate(\"chr1\", 1004500, '-'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7a8f85ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10108</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>18</td>\n",
       "      <td>TAACCC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>10149</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10147</td>\n",
       "      <td>10179</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>32</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10172</td>\n",
       "      <td>10184</td>\n",
       "      <td>AACCT</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10177</td>\n",
       "      <td>10233</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>56</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTAAC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4396267</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363486</td>\n",
       "      <td>59363498</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>59360854</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4396268</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363501</td>\n",
       "      <td>59363513</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>59360854</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4396269</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363514</td>\n",
       "      <td>59363543</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>59360854</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4396270</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363540</td>\n",
       "      <td>59363552</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>59360854</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4396271</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363553</td>\n",
       "      <td>59363565</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>59360854</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2699</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4396272 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | class      | length    | Strand       | num_units   | actual_repeat   | gene       | gene_start   | gene_stop   | gene_strand   | annotation   | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   | (int64)     | (object)        | (object)   | (object)     | (object)    | (object)      | (object)     | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------|\n",
       "| chr1         | 10000     | 10108     | AACCCT     | 108       | +            | 18          | TAACCC          | uc001aaa.3 | 11874        | 14409       | +             | Intergenic   | ...   |\n",
       "| chr1         | 10108     | 10149     | AACCCT     | 41        | +            | 6           | AACCCT          | uc001aaa.3 | 11874        | 14409       | +             | Intergenic   | ...   |\n",
       "| chr1         | 10147     | 10179     | AACCCT     | 32        | +            | 5           | CCCTAA          | uc001aaa.3 | 11874        | 14409       | +             | Intergenic   | ...   |\n",
       "| chr1         | 10172     | 10184     | AACCT      | 12        | +            | 2           | CCTAA           | uc001aaa.3 | 11874        | 14409       | +             | Intergenic   | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...       | ...          | ...         | ...             | ...        | ...          | ...         | ...           | ...          | ...   |\n",
       "| chrY         | 59363501  | 59363513  | ACACCC     | 12        | -            | 2           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "| chrY         | 59363514  | 59363543  | ACACCC     | 29        | -            | 4           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "| chrY         | 59363540  | 59363552  | ACACCC     | 12        | -            | 2           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "| chrY         | 59363553  | 59363565  | ACACCC     | 12        | -            | 2           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------+\n",
       "Stranded PyRanges object has 4,396,272 rows and 15 columns from 92 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: promoter, dist_to_tss"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str_pr = pr.PyRanges(str_df)\n",
    "str_pr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad542a78",
   "metadata": {},
   "source": [
    "Load the UCSC introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "68cca9a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>unk</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12612</td>\n",
       "      <td>uc001aaa.3_intron_0_0_chr1_12228_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13220</td>\n",
       "      <td>uc001aaa.3_intron_1_0_chr1_12722_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12645</td>\n",
       "      <td>uc010nxr.1_intron_0_0_chr1_12228_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12697</td>\n",
       "      <td>13220</td>\n",
       "      <td>uc010nxr.1_intron_1_0_chr1_12698_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12594</td>\n",
       "      <td>uc010nxq.1_intron_0_0_chr1_12228_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End                             ucsc_id  unk Strand\n",
       "0       chr1  12227  12612  uc001aaa.3_intron_0_0_chr1_12228_f    0      +\n",
       "1       chr1  12721  13220  uc001aaa.3_intron_1_0_chr1_12722_f    0      +\n",
       "2       chr1  12227  12645  uc010nxr.1_intron_0_0_chr1_12228_f    0      +\n",
       "3       chr1  12697  13220  uc010nxr.1_intron_1_0_chr1_12698_f    0      +\n",
       "4       chr1  12227  12594  uc010nxq.1_intron_0_0_chr1_12228_f    0      +"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = [\"Chromosome\", \"Start\", \"End\", \"ucsc_id\", \"unk\", \"Strand\"]\n",
    "intron_df = pd.read_csv(\"data/ucsc_introns.bed\", sep = '\\t', names = col_names, index_col = False, skiprows = [0])\n",
    "intron_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d2690c11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>unk</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12612</td>\n",
       "      <td>uc001aaa.3_intron_0_0_chr1_12228_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13220</td>\n",
       "      <td>uc001aaa.3_intron_1_0_chr1_12722_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12645</td>\n",
       "      <td>uc010nxr.1_intron_0_0_chr1_12228_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12697</td>\n",
       "      <td>13220</td>\n",
       "      <td>uc010nxr.1_intron_1_0_chr1_12698_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12594</td>\n",
       "      <td>uc010nxq.1_intron_0_0_chr1_12228_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659322</th>\n",
       "      <td>chrY</td>\n",
       "      <td>27216988</td>\n",
       "      <td>27218792</td>\n",
       "      <td>uc011nbv.2_intron_1_0_chrY_27216989_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659323</th>\n",
       "      <td>chrY</td>\n",
       "      <td>27218868</td>\n",
       "      <td>27245878</td>\n",
       "      <td>uc011nbv.2_intron_2_0_chrY_27218869_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659324</th>\n",
       "      <td>chrY</td>\n",
       "      <td>27329895</td>\n",
       "      <td>27330860</td>\n",
       "      <td>uc004fwt.3_intron_0_0_chrY_27329896_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659325</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59359508</td>\n",
       "      <td>59360006</td>\n",
       "      <td>uc011ncc.1_intron_0_0_chrY_59359509_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659326</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59360115</td>\n",
       "      <td>59360500</td>\n",
       "      <td>uc011ncc.1_intron_1_0_chrY_59360116_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>659327 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+---------------------------------------+-----------+--------------+\n",
       "| Chromosome   | Start     | End       | ucsc_id                               | unk       | Strand       |\n",
       "| (category)   | (int32)   | (int32)   | (object)                              | (int64)   | (category)   |\n",
       "|--------------+-----------+-----------+---------------------------------------+-----------+--------------|\n",
       "| chr1         | 12227     | 12612     | uc001aaa.3_intron_0_0_chr1_12228_f    | 0         | +            |\n",
       "| chr1         | 12721     | 13220     | uc001aaa.3_intron_1_0_chr1_12722_f    | 0         | +            |\n",
       "| chr1         | 12227     | 12645     | uc010nxr.1_intron_0_0_chr1_12228_f    | 0         | +            |\n",
       "| chr1         | 12697     | 13220     | uc010nxr.1_intron_1_0_chr1_12698_f    | 0         | +            |\n",
       "| ...          | ...       | ...       | ...                                   | ...       | ...          |\n",
       "| chrY         | 27218868  | 27245878  | uc011nbv.2_intron_2_0_chrY_27218869_r | 0         | -            |\n",
       "| chrY         | 27329895  | 27330860  | uc004fwt.3_intron_0_0_chrY_27329896_r | 0         | -            |\n",
       "| chrY         | 59359508  | 59360006  | uc011ncc.1_intron_0_0_chrY_59359509_r | 0         | -            |\n",
       "| chrY         | 59360115  | 59360500  | uc011ncc.1_intron_1_0_chrY_59360116_r | 0         | -            |\n",
       "+--------------+-----------+-----------+---------------------------------------+-----------+--------------+\n",
       "Stranded PyRanges object has 659,327 rows and 6 columns from 54 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_pr = pr.PyRanges(intron_df)\n",
    "intron_pr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51504547",
   "metadata": {},
   "source": [
    "Next, get all the intervals of short tandem repeats that are located in introns. The new PyRanges Start, End will now reflect only the region of the repeat that overlapped the intron, not the length of the whole repeat (this information is retained in the `length` column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f7f6a3f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "str_intron_intersect = str_pr.intersect(intron_pr, strandedness= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0cfc00c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15240</td>\n",
       "      <td>15255</td>\n",
       "      <td>AGGGCC</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGCCA</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>14115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15383</td>\n",
       "      <td>15395</td>\n",
       "      <td>AGGCGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCAGGC</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>13975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17476</td>\n",
       "      <td>17488</td>\n",
       "      <td>AGCCG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCGAG</td>\n",
       "      <td>uc009vjc.1</td>\n",
       "      <td>16858</td>\n",
       "      <td>17751</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17801</td>\n",
       "      <td>17814</td>\n",
       "      <td>ATCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCATC</td>\n",
       "      <td>uc009vjd.2</td>\n",
       "      <td>15796</td>\n",
       "      <td>18061</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18453</td>\n",
       "      <td>18466</td>\n",
       "      <td>AGGCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCAG</td>\n",
       "      <td>uc009vit.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>19759</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1973757</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59342100</td>\n",
       "      <td>59342121</td>\n",
       "      <td>AC</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>10</td>\n",
       "      <td>TG</td>\n",
       "      <td>uc004fxm.1</td>\n",
       "      <td>59330252</td>\n",
       "      <td>59343488</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1973758</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59354264</td>\n",
       "      <td>59354277</td>\n",
       "      <td>AGGCC</td>\n",
       "      <td>13</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GGCCT</td>\n",
       "      <td>uc004fxo.1</td>\n",
       "      <td>59352973</td>\n",
       "      <td>59356131</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1973759</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59355249</td>\n",
       "      <td>59355261</td>\n",
       "      <td>AGCCG</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GGCTC</td>\n",
       "      <td>uc022cpg.1</td>\n",
       "      <td>59354985</td>\n",
       "      <td>59358336</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1973760</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59357345</td>\n",
       "      <td>59357357</td>\n",
       "      <td>AGGCGC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCTGC</td>\n",
       "      <td>uc022cpg.1</td>\n",
       "      <td>59354985</td>\n",
       "      <td>59358336</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>2360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1973761</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59357485</td>\n",
       "      <td>59357500</td>\n",
       "      <td>AGGGCC</td>\n",
       "      <td>15</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCTGG</td>\n",
       "      <td>uc022cpg.1</td>\n",
       "      <td>59354985</td>\n",
       "      <td>59358336</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>2500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1973762 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | class      | length    | Strand       | num_units   | actual_repeat   | gene       | gene_start   | gene_stop   | gene_strand   | annotation   | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   | (int64)     | (object)        | (object)   | (object)     | (object)    | (object)      | (object)     | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------|\n",
       "| chr1         | 15240     | 15255     | AGGGCC     | 15        | +            | 2           | GGGCCA          | uc009viv.2 | 14407        | 29370       | -             | Exon         | ...   |\n",
       "| chr1         | 15383     | 15395     | AGGCGC     | 12        | +            | 2           | GCAGGC          | uc009viv.2 | 14407        | 29370       | -             | Exon         | ...   |\n",
       "| chr1         | 17476     | 17488     | AGCCG      | 12        | +            | 2           | CCGAG           | uc009vjc.1 | 16858        | 17751       | -             | Exon         | ...   |\n",
       "| chr1         | 17801     | 17814     | ATCCC      | 13        | +            | 2           | CCATC           | uc009vjd.2 | 15796        | 18061       | -             | Exon         | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...       | ...          | ...         | ...             | ...        | ...          | ...         | ...           | ...          | ...   |\n",
       "| chrY         | 59354264  | 59354277  | AGGCC      | 13        | -            | 2           | GGCCT           | uc004fxo.1 | 59352973     | 59356131    | +             | Intron       | ...   |\n",
       "| chrY         | 59355249  | 59355261  | AGCCG      | 12        | -            | 2           | GGCTC           | uc022cpg.1 | 59354985     | 59358336    | +             | Intron       | ...   |\n",
       "| chrY         | 59357345  | 59357357  | AGGCGC     | 12        | -            | 2           | GCCTGC          | uc022cpg.1 | 59354985     | 59358336    | +             | Intron       | ...   |\n",
       "| chrY         | 59357485  | 59357500  | AGGGCC     | 15        | -            | 2           | CCCTGG          | uc022cpg.1 | 59354985     | 59358336    | +             | Intron       | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------+\n",
       "Stranded PyRanges object has 1,973,762 rows and 15 columns from 53 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: promoter, dist_to_tss"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str_intron_intersect\n",
    "# drop duplicate rows in the intersection PyRanges\n",
    "str_intron_intersect_nodup = str_intron_intersect.drop_duplicate_positions()\n",
    "str_intron_intersect_nodup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8342fd8b",
   "metadata": {},
   "source": [
    "Now, get the STRs that are within 300kb of an intron. First, find the nearest intron to each STR:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "89154bf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>unk</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15240</td>\n",
       "      <td>15255</td>\n",
       "      <td>AGGGCC</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGCCA</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>14115</td>\n",
       "      <td>14829</td>\n",
       "      <td>15795</td>\n",
       "      <td>uc009viq.3_intron_0_0_chr1_14830_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15383</td>\n",
       "      <td>15395</td>\n",
       "      <td>AGGCGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCAGGC</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>13975</td>\n",
       "      <td>14829</td>\n",
       "      <td>15795</td>\n",
       "      <td>uc009viq.3_intron_0_0_chr1_14830_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17476</td>\n",
       "      <td>17488</td>\n",
       "      <td>AGCCG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCGAG</td>\n",
       "      <td>uc009vjc.1</td>\n",
       "      <td>16858</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>263</td>\n",
       "      <td>17055</td>\n",
       "      <td>17605</td>\n",
       "      <td>uc009viq.3_intron_3_0_chr1_17056_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17801</td>\n",
       "      <td>17814</td>\n",
       "      <td>ATCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCATC</td>\n",
       "      <td>uc009vjd.2</td>\n",
       "      <td>15796</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>247</td>\n",
       "      <td>17368</td>\n",
       "      <td>17914</td>\n",
       "      <td>uc009vix.2_intron_3_0_chr1_17369_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18453</td>\n",
       "      <td>18466</td>\n",
       "      <td>AGGCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCAG</td>\n",
       "      <td>uc009vit.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1293</td>\n",
       "      <td>18061</td>\n",
       "      <td>24737</td>\n",
       "      <td>uc009vix.2_intron_4_0_chr1_18062_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4390671</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363486</td>\n",
       "      <td>59363498</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2632</td>\n",
       "      <td>59360115</td>\n",
       "      <td>59360500</td>\n",
       "      <td>uc011ncc.1_intron_1_0_chrY_59360116_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>2987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4390672</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363501</td>\n",
       "      <td>59363513</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2647</td>\n",
       "      <td>59360115</td>\n",
       "      <td>59360500</td>\n",
       "      <td>uc011ncc.1_intron_1_0_chrY_59360116_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>3002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4390673</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363514</td>\n",
       "      <td>59363543</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2660</td>\n",
       "      <td>59360115</td>\n",
       "      <td>59360500</td>\n",
       "      <td>uc011ncc.1_intron_1_0_chrY_59360116_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>3015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4390674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363540</td>\n",
       "      <td>59363552</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2686</td>\n",
       "      <td>59360115</td>\n",
       "      <td>59360500</td>\n",
       "      <td>uc011ncc.1_intron_1_0_chrY_59360116_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>3041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4390675</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59363553</td>\n",
       "      <td>59363565</td>\n",
       "      <td>ACACCC</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGTGG</td>\n",
       "      <td>uc011ncc.1</td>\n",
       "      <td>59358329</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-2699</td>\n",
       "      <td>59360115</td>\n",
       "      <td>59360500</td>\n",
       "      <td>uc011ncc.1_intron_1_0_chrY_59360116_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>3054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4390676 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | class      | length    | Strand       | num_units   | actual_repeat   | gene       | gene_start   | gene_stop   | gene_strand   | annotation   | +8    |\n",
       "| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   | (int64)     | (object)        | (object)   | (object)     | (object)    | (object)      | (object)     | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------|\n",
       "| chr1         | 15240     | 15255     | AGGGCC     | 15        | +            | 2           | GGGCCA          | uc009viv.2 | 14407        | 29370       | -             | Exon         | ...   |\n",
       "| chr1         | 15383     | 15395     | AGGCGC     | 12        | +            | 2           | GCAGGC          | uc009viv.2 | 14407        | 29370       | -             | Exon         | ...   |\n",
       "| chr1         | 17476     | 17488     | AGCCG      | 12        | +            | 2           | CCGAG           | uc009vjc.1 | 16858        | 17751       | -             | Exon         | ...   |\n",
       "| chr1         | 17801     | 17814     | ATCCC      | 13        | +            | 2           | CCATC           | uc009vjd.2 | 15796        | 18061       | -             | Exon         | ...   |\n",
       "| ...          | ...       | ...       | ...        | ...       | ...          | ...         | ...             | ...        | ...          | ...         | ...           | ...          | ...   |\n",
       "| chrY         | 59363501  | 59363513  | ACACCC     | 12        | -            | 2           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "| chrY         | 59363514  | 59363543  | ACACCC     | 29        | -            | 4           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "| chrY         | 59363540  | 59363552  | ACACCC     | 12        | -            | 2           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "| chrY         | 59363553  | 59363565  | ACACCC     | 12        | -            | 2           | GTGTGG          | uc011ncc.1 | 59358329     | 59360854    | -             | Intergenic   | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-----------------+------------+--------------+-------------+---------------+--------------+-------+\n",
       "Stranded PyRanges object has 4,390,676 rows and 21 columns from 54 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: promoter, dist_to_tss, Start_b, End_b, ucsc_id, unk, Strand_b, Distance"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_intron_pr = str_pr.nearest(intron_pr)\n",
    "nearest_intron_pr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "330cb9b2",
   "metadata": {},
   "source": [
    "Then, get everything "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64e59f21",
   "metadata": {},
   "source": [
    "Next, want to get the eCLIP peaks that are in these STRs. To start, let's try "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc-autonumbering": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
