{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4570e66e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import time\n",
    "import json\n",
    "# Display all the columns of dataframes\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c3a151",
   "metadata": {},
   "source": [
    "# Load STR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b34398e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>Length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13623</td>\n",
       "      <td>13635</td>\n",
       "      <td>ACAGGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CAGGCA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1749</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>uc010nxr.1</td>\n",
       "      <td>uc010nxr.1</td>\n",
       "      <td>3</td>\n",
       "      <td>uc010nxr.1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>322376</td>\n",
       "      <td>322388</td>\n",
       "      <td>AAGAAT</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GAATAA</td>\n",
       "      <td>uc009vjk.2</td>\n",
       "      <td>322037</td>\n",
       "      <td>326938</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>339</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>322036</td>\n",
       "      <td>322228</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>C9J4L2</td>\n",
       "      <td>uc009vjk.2</td>\n",
       "      <td>1</td>\n",
       "      <td>uc009vjk.2.1</td>\n",
       "      <td>C9J4L2</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>877247</td>\n",
       "      <td>877259</td>\n",
       "      <td>AGGGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGGG</td>\n",
       "      <td>uc001abx.1</td>\n",
       "      <td>871152</td>\n",
       "      <td>879961</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>6095</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>877515</td>\n",
       "      <td>877631</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>H7BY14</td>\n",
       "      <td>uc031pjy.1</td>\n",
       "      <td>8</td>\n",
       "      <td>uc031pjy.1.8</td>\n",
       "      <td>H7BY14</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879886</td>\n",
       "      <td>879906</td>\n",
       "      <td>AGCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>CCAGC</td>\n",
       "      <td>uc001abx.1</td>\n",
       "      <td>871152</td>\n",
       "      <td>879961</td>\n",
       "      <td>+</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>8734</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>879287</td>\n",
       "      <td>879961</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Q96NU1</td>\n",
       "      <td>uc031pjt.1</td>\n",
       "      <td>12</td>\n",
       "      <td>uc031pjt.1.12</td>\n",
       "      <td>Q96NU1</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>882085</td>\n",
       "      <td>882097</td>\n",
       "      <td>AGAGGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGCAGA</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11821</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>881781</td>\n",
       "      <td>882440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>uc031pkq.1</td>\n",
       "      <td>5</td>\n",
       "      <td>uc031pkq.1.5</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>882362</td>\n",
       "      <td>882374</td>\n",
       "      <td>AGATGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGAGAT</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11544</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>881781</td>\n",
       "      <td>882440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>uc031pkq.1</td>\n",
       "      <td>5</td>\n",
       "      <td>uc031pkq.1.5</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>882442</td>\n",
       "      <td>882456</td>\n",
       "      <td>ACT</td>\n",
       "      <td>14</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>TAC</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11462</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>881781</td>\n",
       "      <td>882440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>uc031pkq.1</td>\n",
       "      <td>5</td>\n",
       "      <td>uc031pkq.1.5</td>\n",
       "      <td>Q8NAK7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>888115</td>\n",
       "      <td>888127</td>\n",
       "      <td>ACCCCC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCCA</td>\n",
       "      <td>uc001aby.4</td>\n",
       "      <td>879583</td>\n",
       "      <td>893918</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>5791</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>887791</td>\n",
       "      <td>887980</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NP_056473</td>\n",
       "      <td>uc001abz.4</td>\n",
       "      <td>10</td>\n",
       "      <td>uc001abz.4.10</td>\n",
       "      <td>NP_056473</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    Length | +23   |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+-------|\n",
       "| chr1         |     13623 |     13635 | ACAGGC     |        12 | ...   |\n",
       "| chr1         |    322376 |    322388 | AAGAAT     |        12 | ...   |\n",
       "| chr1         |    877247 |    877259 | AGGGG      |        12 | ...   |\n",
       "| chr1         |    879886 |    879906 | AGCCC      |        20 | ...   |\n",
       "| chr1         |    882085 |    882097 | AGAGGC     |        12 | ...   |\n",
       "| chr1         |    882362 |    882374 | AGATGG     |        12 | ...   |\n",
       "| chr1         |    882442 |    882456 | ACT        |        14 | ...   |\n",
       "| chr1         |    888115 |    888127 | ACCCCC     |        12 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 28 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "23 hidden columns: Strand, num_units, actual_repeat, gene_id, gene_start, gene_stop, ... (+ 17 more.)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "within_300 = pr.PyRanges(pd.read_csv(\"data/temp/within_300_ss_pyranges.csv\"))\n",
    "within_300.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3613da62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>Length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>22811</td>\n",
       "      <td>22823</td>\n",
       "      <td>AAAGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGAA</td>\n",
       "      <td>uc001aac.4</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>6547</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>24737</td>\n",
       "      <td>24891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>9</td>\n",
       "      <td>uc009vir.3.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23874</td>\n",
       "      <td>23886</td>\n",
       "      <td>ACCTCC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCTCCA</td>\n",
       "      <td>uc009viq.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>5484</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>24737</td>\n",
       "      <td>24891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>uc009vir.3</td>\n",
       "      <td>9</td>\n",
       "      <td>uc009vir.3.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>669727</td>\n",
       "      <td>669739</td>\n",
       "      <td>AAC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>AAC</td>\n",
       "      <td>uc009vjm.3</td>\n",
       "      <td>661139</td>\n",
       "      <td>670994</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1255</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>668686</td>\n",
       "      <td>668744</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>B4DGA0</td>\n",
       "      <td>uc009vjm.3</td>\n",
       "      <td>2</td>\n",
       "      <td>uc009vjm.3.2</td>\n",
       "      <td>B4DGA0</td>\n",
       "      <td>984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>676490</td>\n",
       "      <td>676503</td>\n",
       "      <td>AAAAG</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AAAAG</td>\n",
       "      <td>uc001abm.2</td>\n",
       "      <td>674240</td>\n",
       "      <td>679736</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>3233</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>675508</td>\n",
       "      <td>675566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abm.2</td>\n",
       "      <td>uc001abm.2</td>\n",
       "      <td>3</td>\n",
       "      <td>uc001abm.2.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>700958</td>\n",
       "      <td>700970</td>\n",
       "      <td>AAATAC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>TACAAA</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>13098</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>700244</td>\n",
       "      <td>700627</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>1</td>\n",
       "      <td>uc001abo.3.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>701123</td>\n",
       "      <td>701136</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>A</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>12932</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>700244</td>\n",
       "      <td>700627</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>1</td>\n",
       "      <td>uc001abo.3.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>701240</td>\n",
       "      <td>701253</td>\n",
       "      <td>ACTCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCACT</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>12815</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>701708</td>\n",
       "      <td>701767</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>2</td>\n",
       "      <td>uc001abo.3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>702159</td>\n",
       "      <td>702173</td>\n",
       "      <td>AATC</td>\n",
       "      <td>14</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>ATCA</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>700245</td>\n",
       "      <td>714068</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>11895</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>701708</td>\n",
       "      <td>701767</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>uc001abo.3</td>\n",
       "      <td>2</td>\n",
       "      <td>uc001abo.3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    Length | +23   |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+-------|\n",
       "| chr1         |     22811 |     22823 | AAAGG      |        12 | ...   |\n",
       "| chr1         |     23874 |     23886 | ACCTCC     |        12 | ...   |\n",
       "| chr1         |    669727 |    669739 | AAC        |        12 | ...   |\n",
       "| chr1         |    676490 |    676503 | AAAAG      |        13 | ...   |\n",
       "| chr1         |    700958 |    700970 | AAATAC     |        12 | ...   |\n",
       "| chr1         |    701123 |    701136 | A          |        13 | ...   |\n",
       "| chr1         |    701240 |    701253 | ACTCC      |        13 | ...   |\n",
       "| chr1         |    702159 |    702173 | AATC       |        14 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+-------+\n",
       "Stranded PyRanges object has 8 rows and 28 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "23 hidden columns: Strand, num_units, actual_repeat, gene_id, gene_start, gene_stop, ... (+ 17 more.)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_internal = pr.PyRanges(pd.read_csv(\"data/temp/intron_internal_pyranges.csv\"))\n",
    "intron_internal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b01a31c",
   "metadata": {},
   "source": [
    "# Compare each STR pyranges to all the eCLIP peaks pyranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83f8991f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# prune some of the columns from our pyranges\n",
    "columns = ['Chromosome', 'Start', 'End', 'class', 'Length', 'Strand', 'num_units', 'actual_repeat','gene_id',\n",
    "           'gene_strand','annotation','promoter','dist_to_tss','Distance']\n",
    "\n",
    "eclip_STR_internal_df = intron_internal.as_df()[columns]\n",
    "eclip_STR_internal_df = eclip_STR_internal_df.rename(columns = {'Distance': 'dist_to_ss', 'annotation':'str_annotation', 'promoter':'str_in_promoter'})\n",
    "\n",
    "eclip_STR_within_300_df = within_300.as_df()[columns]\n",
    "eclip_STR_within_300_df = eclip_STR_within_300_df.rename(columns = {'Distance': 'dist_to_ss'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7af548aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.9527838349342346\n",
      "20 done, 398.3679023265839s remaining\n",
      "1.9843619227409364\n",
      "40 done, 365.1225937843323s remaining\n",
      "2.0076157649358115\n",
      "60 done, 329.2489854494731s remaining\n",
      "2.0085935831069945\n",
      "80 done, 289.2374759674072s remaining\n",
      "2.00246869802475\n",
      "100 done, 248.306118555069s remaining\n",
      "2.052915006875992\n",
      "120 done, 213.50316071510315s remaining\n",
      "2.072046768665314\n",
      "140 done, 174.05192856788636s remaining\n",
      "2.0703290343284606\n",
      "160 done, 132.50105819702148s remaining\n",
      "2.0772620452774895\n",
      "180 done, 91.39952999220954s remaining\n",
      "2.1237760019302367\n",
      "200 done, 50.97062404632568s remaining\n",
      "2.1107252316041425\n",
      "220 done, 8.44290092641657s remaining\n"
     ]
    }
   ],
   "source": [
    "times = np.array([])\n",
    "i = 0\n",
    "# all_eCLIP_dict = {}\n",
    "eclip_id2rbp = {}\n",
    "\n",
    "for filename in os.listdir(\"data/eclip/\"):\n",
    "    start = time.time()\n",
    "    if filename.endswith(\".bed\"):\n",
    "        dataset_name = filename.split(\".\")[0]\n",
    "        current_pr = pr.read_bed(\"data/eclip/{}\".format(filename))\n",
    "        rbp_name = current_pr.as_df().iloc[0,3].rstrip('_IDR')\n",
    "        # all_eCLIP_dict[dataset_name] = current_pr\n",
    "        overlap_count = intron_internal.count_overlaps(current_pr, \n",
    "                                                       overlap_col = dataset_name, \n",
    "                                                       strandedness = False)\n",
    "        eclip_id2rbp[dataset_name] = rbp_name\n",
    "        col_to_add = overlap_count.__getattr__(dataset_name)\n",
    "        eclip_STR_internal_df[dataset_name] = np.array(col_to_add)\n",
    "\n",
    "    end = time.time()\n",
    "    times = np.append(times, end-start)\n",
    "    i+=1\n",
    "    if i % 20 == 0:\n",
    "        mean_time = np.mean(times)\n",
    "        print(mean_time)\n",
    "        print(\"{0} done, {1}s remaining\".format(i, mean_time*(len(os.listdir(\"data/eclip/\"))-i)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ceb859f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "num_eclip_peaks = eclip_STR_internal_df.iloc[:,14:].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "dcc3baa6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PTBP1_HepG2\n",
      "PTBP1_K562\n",
      "HNRNPL_HepG2\n",
      "PCBP2_HepG2\n",
      "TARDBP_K562\n",
      "HNRNPL_K562\n",
      "EIF4G2_K562\n",
      "AGGF1_K562\n",
      "CSTF2T_HepG2\n",
      "AGGF1_HepG2\n"
     ]
    }
   ],
   "source": [
    "for rbp in num_eclip_peaks.sort_values(ascending=False)[:10].index.values:\n",
    "    print(eclip_id2rbp[rbp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "179f6df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip_STR_internal_df.to_csv(\"data/temp/eclip_STR_internal_df.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1ee785c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9334316968917846\n",
      "20 done, 190.42006616592406s remaining\n",
      "0.9222236394882202\n",
      "40 done, 169.6891496658325s remaining\n",
      "0.9059597412745158\n",
      "60 done, 148.5773975690206s remaining\n",
      "0.8930589854717255\n",
      "80 done, 128.60049390792847s remaining\n",
      "0.8957520484924316\n",
      "100 done, 111.07325401306151s remaining\n",
      "0.8885562757651011\n",
      "120 done, 92.40985267957052s remaining\n",
      "0.8891740475382124\n",
      "140 done, 74.69061999320984s remaining\n",
      "0.8909909963607788\n",
      "160 done, 57.023423767089845s remaining\n",
      "0.8981874439451429\n",
      "180 done, 39.520247533586286s remaining\n",
      "0.903650255203247\n",
      "200 done, 21.68760612487793s remaining\n",
      "0.9027645826339722\n",
      "220 done, 3.6110583305358888s remaining\n"
     ]
    }
   ],
   "source": [
    "times = np.array([])\n",
    "i = 0\n",
    "for filename in os.listdir(\"data/eclip/\"):\n",
    "    start = time.time()\n",
    "    if filename.endswith(\".bed\"):\n",
    "        dataset_name = filename.split(\".\")[0]\n",
    "        current_pr = pr.read_bed(\"data/eclip/{}\".format(filename))\n",
    "        rbp_name = current_pr.as_df().iloc[0,3].rstrip('_IDR')\n",
    "        overlap_count = within_300.count_overlaps(current_pr, \n",
    "                                                       overlap_col = dataset_name, \n",
    "                                                       strandedness = False)\n",
    "        col_to_add = overlap_count.__getattr__(dataset_name)\n",
    "        eclip_STR_within_300_df[dataset_name] = np.array(col_to_add)\n",
    "\n",
    "    end = time.time()\n",
    "    times = np.append(times, end-start)\n",
    "    i+=1\n",
    "    if i % 20 == 0:\n",
    "        mean_time = np.mean(times)\n",
    "        print(mean_time)\n",
    "        print(\"{0} done, {1}s remaining\".format(i, mean_time*(len(os.listdir(\"data/eclip/\"))-i)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2751a95c",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_eclip_peaks = eclip_STR_within_300_df.iloc[:,14:].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3ef10f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DDX3X_HepG2\n",
      "UCHL5_K562\n",
      "AQR_K562\n",
      "PRPF8_HepG2\n",
      "BCLAF1_HepG2\n",
      "GRWD1_HepG2\n",
      "PPIG_HepG2\n",
      "ZNF622_K562\n",
      "PRPF4_HepG2\n",
      "AQR_HepG2\n"
     ]
    }
   ],
   "source": [
    "for rbp in num_eclip_peaks.sort_values(ascending=False)[:10].index.values:\n",
    "    print(eclip_id2rbp[rbp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6d6cd1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip_STR_within_300_df.to_csv(\"data/temp/eclip_STR_within_300_df.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "37fee57e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/temp/encode_id2rbp.json\", \"w\") as file:\n",
    "    json.dump(eclip_id2rbp, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "af4ddc28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chr7', 'chr22', 'chr3', 'chr5', 'chr14', 'chr15', 'chrM', 'chr17', 'chr20', 'chr16', 'chr8', 'chr21', 'chr19', 'chrY', 'chr2', 'chr1', 'chr4', 'chr6', 'chr9', 'chr12', 'chr18', 'chr11', 'chr13', 'chrX', 'chr10'}\n"
     ]
    }
   ],
   "source": [
    "# filter out the weird/unmapped chromosome fragments in intron_internal and within_300\n",
    "# eCLIP_chr = set()\n",
    "# for key in all_eCLIP_dict.keys():\n",
    "#     b = set(all_eCLIP_dict[key].Chromosome)\n",
    "#     eCLIP_chr=eCLIP_chr.union(b)\n",
    "# print(eCLIP_chr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "333b0d16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# intron_internal = intron_internal[intron_internal.Chromosome.isin(eCLIP_chr)]\n",
    "# within_300 = within_300[within_300.Chromosome.isin(eCLIP_chr)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
