{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d9385667",
   "metadata": {},
   "source": [
    "# Identify STRs that are at splice sites\n",
    "Like the ENCODE paper, define a splice site as being within 100bp of an exon. All of the following uses the GRCh37 alignments (GENCODE v19 annotations), also like the paper. \n",
    "Links to [GENCODE data](https://www.gencodegenes.org/human/release_37lift37.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "16db1749",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bd8aa21",
   "metadata": {},
   "source": [
    "Load the STR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82839d14",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/david/Library/Python/3.7/lib/python/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (9,10,14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10108</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>18</td>\n",
       "      <td>TAACCC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>10149</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10147</td>\n",
       "      <td>10179</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>32</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10172</td>\n",
       "      <td>10184</td>\n",
       "      <td>AACCT</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10177</td>\n",
       "      <td>10233</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>56</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTAAC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End   class  length Strand  num_units actual_repeat  \\\n",
       "0       chr1  10000  10108  AACCCT     108      +         18        TAACCC   \n",
       "1       chr1  10108  10149  AACCCT      41      +          6        AACCCT   \n",
       "2       chr1  10147  10179  AACCCT      32      +          5        CCCTAA   \n",
       "3       chr1  10172  10184   AACCT      12      +          2         CCTAA   \n",
       "4       chr1  10177  10233  AACCCT      56      +          9        CCTAAC   \n",
       "\n",
       "         gene gene_start gene_stop gene_strand  annotation      promoter  \\\n",
       "0  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "1  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "2  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "3  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "4  uc001aaa.3      11874     14409           +  Intergenic  Non-Promoter   \n",
       "\n",
       "  dist_to_tss  \n",
       "0       -1766  \n",
       "1       -1725  \n",
       "2       -1695  \n",
       "3       -1690  \n",
       "4       -1641  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = [\"Chromosome\", \"Start\", \"End\", \"class\", \n",
    "            \"length\", \"Strand\", \"num_units\", \n",
    "            \"actual_repeat\", \"gene\", \"gene_start\",\n",
    "            \"gene_stop\", \"gene_strand\", \"annotation\",\n",
    "            \"promoter\", \"dist_to_tss\"]\n",
    "dtype_dict = {\"Chromosome\": np.str, \"Start\":np.int, \"End\":np.int, \"length\": np.int, \n",
    "             \"Strand\": np.str} #, \"gene_start\": np.int, \"gene_stop\": np.int, \"dist_to_tss\": np.int}\n",
    "str_df = pd.read_csv(\"data/msdb_data.tsv\", sep = '\\t', dtype = dtype_dict, names = col_names, index_col = False, na_values = \"\")\n",
    "str_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a7dfda4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10108</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>18</td>\n",
       "      <td>TAACCC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>10149</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10147</td>\n",
       "      <td>10179</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>32</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10172</td>\n",
       "      <td>10184</td>\n",
       "      <td>AACCT</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10177</td>\n",
       "      <td>10233</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>56</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTAAC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10231</td>\n",
       "      <td>10249</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>18</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>CCCTAA</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10255</td>\n",
       "      <td>10290</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>35</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>AACCCT</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10285</td>\n",
       "      <td>10326</td>\n",
       "      <td>AACCCC</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>AACCCC</td>\n",
       "      <td>uc001aaa.3</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>-1548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    length | Strand       |   num_units | +8    |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |     (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+--------------+-------------+-------|\n",
       "| chr1         |     10000 |     10108 | AACCCT     |       108 | +            |          18 | ...   |\n",
       "| chr1         |     10108 |     10149 | AACCCT     |        41 | +            |           6 | ...   |\n",
       "| chr1         |     10147 |     10179 | AACCCT     |        32 | +            |           5 | ...   |\n",
       "| chr1         |     10172 |     10184 | AACCT      |        12 | +            |           2 | ...   |\n",
       "| chr1         |     10177 |     10233 | AACCCT     |        56 | +            |           9 | ...   |\n",
       "| chr1         |     10231 |     10249 | AACCCT     |        18 | +            |           3 | ...   |\n",
       "| chr1         |     10255 |     10290 | AACCCT     |        35 | +            |           5 | ...   |\n",
       "| chr1         |     10285 |     10326 | AACCCC     |        41 | +            |           6 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 15 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: actual_repeat, gene, gene_start, gene_stop, gene_strand, annotation, promoter, dist_to_tss"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str_pr = pr.PyRanges(str_df)\n",
    "str_pr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cf6345c",
   "metadata": {},
   "source": [
    "Load the UCSC exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8dfd560b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>12227</td>\n",
       "      <td>uc001aaa.3_exon_0_0_chr1_11874_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>uc001aaa.3_exon_1_0_chr1_12613_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>uc001aaa.3_exon_2_0_chr1_13221_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>12227</td>\n",
       "      <td>uc010nxr.1_exon_0_0_chr1_11874_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12645</td>\n",
       "      <td>12697</td>\n",
       "      <td>uc010nxr.1_exon_1_0_chr1_12646_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End                           ucsc_id Strand\n",
       "0       chr1  11873  12227  uc001aaa.3_exon_0_0_chr1_11874_f      +\n",
       "1       chr1  12612  12721  uc001aaa.3_exon_1_0_chr1_12613_f      +\n",
       "2       chr1  13220  14409  uc001aaa.3_exon_2_0_chr1_13221_f      +\n",
       "3       chr1  11873  12227  uc010nxr.1_exon_0_0_chr1_11874_f      +\n",
       "4       chr1  12645  12697  uc010nxr.1_exon_1_0_chr1_12646_f      +"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = [\"Chromosome\", \"Start\", \"End\", \"ucsc_id\", \"Strand\"]\n",
    "exon_df = pd.read_csv(\"data/ucsc_exons.bed\", sep = '\\t', names = col_names, index_col = False, skiprows = [0], usecols = [0,1,2,3,5])\n",
    "exon_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f48024cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>12227</td>\n",
       "      <td>uc001aaa.3_exon_0_0_chr1_11874_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>uc001aaa.3_exon_1_0_chr1_12613_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>uc001aaa.3_exon_2_0_chr1_13221_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>12227</td>\n",
       "      <td>uc010nxr.1_exon_0_0_chr1_11874_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12645</td>\n",
       "      <td>12697</td>\n",
       "      <td>uc010nxr.1_exon_1_0_chr1_12646_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>uc010nxr.1_exon_2_0_chr1_13221_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>12227</td>\n",
       "      <td>uc010nxq.1_exon_0_0_chr1_11874_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12594</td>\n",
       "      <td>12721</td>\n",
       "      <td>uc010nxq.1_exon_1_0_chr1_12595_f</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+----------------------------------+--------------+\n",
       "| Chromosome   |     Start |       End | ucsc_id                          | Strand       |\n",
       "| (category)   |   (int32) |   (int32) | (object)                         | (category)   |\n",
       "|--------------+-----------+-----------+----------------------------------+--------------|\n",
       "| chr1         |     11873 |     12227 | uc001aaa.3_exon_0_0_chr1_11874_f | +            |\n",
       "| chr1         |     12612 |     12721 | uc001aaa.3_exon_1_0_chr1_12613_f | +            |\n",
       "| chr1         |     13220 |     14409 | uc001aaa.3_exon_2_0_chr1_13221_f | +            |\n",
       "| chr1         |     11873 |     12227 | uc010nxr.1_exon_0_0_chr1_11874_f | +            |\n",
       "| chr1         |     12645 |     12697 | uc010nxr.1_exon_1_0_chr1_12646_f | +            |\n",
       "| chr1         |     13220 |     14409 | uc010nxr.1_exon_2_0_chr1_13221_f | +            |\n",
       "| chr1         |     11873 |     12227 | uc010nxq.1_exon_0_0_chr1_11874_f | +            |\n",
       "| chr1         |     12594 |     12721 | uc010nxq.1_exon_1_0_chr1_12595_f | +            |\n",
       "+--------------+-----------+-----------+----------------------------------+--------------+\n",
       "Stranded PyRanges object has 8 rows and 5 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_pr = pr.PyRanges(exon_df)\n",
    "exon_pr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44835c91",
   "metadata": {},
   "source": [
    "Next, get all the intervals of short tandem repeats that are located in introns. The new PyRanges Start, End will now reflect only the region of the repeat that overlapped the intron, not the length of the whole repeat (this information is retained in the `length` column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6d4e47b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "str_intron_intersect = str_pr.intersect(intron_pr, strandedness= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4d57a6cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15240</td>\n",
       "      <td>15255</td>\n",
       "      <td>AGGGCC</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGCCA</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>14115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15383</td>\n",
       "      <td>15395</td>\n",
       "      <td>AGGCGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCAGGC</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>13975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17476</td>\n",
       "      <td>17488</td>\n",
       "      <td>AGCCG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCGAG</td>\n",
       "      <td>uc009vjc.1</td>\n",
       "      <td>16858</td>\n",
       "      <td>17751</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17801</td>\n",
       "      <td>17814</td>\n",
       "      <td>ATCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCATC</td>\n",
       "      <td>uc009vjd.2</td>\n",
       "      <td>15796</td>\n",
       "      <td>18061</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18453</td>\n",
       "      <td>18466</td>\n",
       "      <td>AGGCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCAG</td>\n",
       "      <td>uc009vit.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>19759</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19412</td>\n",
       "      <td>19424</td>\n",
       "      <td>AGGGGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGAGG</td>\n",
       "      <td>uc009vit.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>19759</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19516</td>\n",
       "      <td>19528</td>\n",
       "      <td>AAAGCC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AAGCCA</td>\n",
       "      <td>uc001aai.1</td>\n",
       "      <td>16858</td>\n",
       "      <td>19759</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>22811</td>\n",
       "      <td>22823</td>\n",
       "      <td>AAAGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGAA</td>\n",
       "      <td>uc001aac.4</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>6547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    length | Strand       |   num_units | +8    |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |     (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+--------------+-------------+-------|\n",
       "| chr1         |     15240 |     15255 | AGGGCC     |        15 | +            |           2 | ...   |\n",
       "| chr1         |     15383 |     15395 | AGGCGC     |        12 | +            |           2 | ...   |\n",
       "| chr1         |     17476 |     17488 | AGCCG      |        12 | +            |           2 | ...   |\n",
       "| chr1         |     17801 |     17814 | ATCCC      |        13 | +            |           2 | ...   |\n",
       "| chr1         |     18453 |     18466 | AGGCC      |        13 | +            |           2 | ...   |\n",
       "| chr1         |     19412 |     19424 | AGGGGG     |        12 | +            |           2 | ...   |\n",
       "| chr1         |     19516 |     19528 | AAAGCC     |        12 | +            |           2 | ...   |\n",
       "| chr1         |     22811 |     22823 | AAAGG      |        12 | +            |           2 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 15 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "8 hidden columns: actual_repeat, gene, gene_start, gene_stop, gene_strand, annotation, promoter, dist_to_tss"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop duplicate rows in the intersection PyRanges\n",
    "str_intron_intersect_nodup = str_intron_intersect.drop_duplicate_positions()\n",
    "str_intron_intersect_nodup.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16990582",
   "metadata": {},
   "source": [
    "Now, get the STRs that are within 300kb of an intron. First, find the nearest intron to each STR:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a9499f2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>Strand</th>\n",
       "      <th>num_units</th>\n",
       "      <th>actual_repeat</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>promoter</th>\n",
       "      <th>dist_to_tss</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>unk</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15240</td>\n",
       "      <td>15255</td>\n",
       "      <td>AGGGCC</td>\n",
       "      <td>15</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGCCA</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>14115</td>\n",
       "      <td>14829</td>\n",
       "      <td>15795</td>\n",
       "      <td>uc009viq.3_intron_0_0_chr1_14830_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15383</td>\n",
       "      <td>15395</td>\n",
       "      <td>AGGCGC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCAGGC</td>\n",
       "      <td>uc009viv.2</td>\n",
       "      <td>14407</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>13975</td>\n",
       "      <td>14829</td>\n",
       "      <td>15795</td>\n",
       "      <td>uc009viq.3_intron_0_0_chr1_14830_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17476</td>\n",
       "      <td>17488</td>\n",
       "      <td>AGCCG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCGAG</td>\n",
       "      <td>uc009vjc.1</td>\n",
       "      <td>16858</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>263</td>\n",
       "      <td>17055</td>\n",
       "      <td>17605</td>\n",
       "      <td>uc009viq.3_intron_3_0_chr1_17056_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17801</td>\n",
       "      <td>17814</td>\n",
       "      <td>ATCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CCATC</td>\n",
       "      <td>uc009vjd.2</td>\n",
       "      <td>15796</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>247</td>\n",
       "      <td>17368</td>\n",
       "      <td>17914</td>\n",
       "      <td>uc009vix.2_intron_3_0_chr1_17369_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18453</td>\n",
       "      <td>18466</td>\n",
       "      <td>AGGCC</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCAG</td>\n",
       "      <td>uc009vit.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>1293</td>\n",
       "      <td>18061</td>\n",
       "      <td>24737</td>\n",
       "      <td>uc009vix.2_intron_4_0_chr1_18062_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19412</td>\n",
       "      <td>19424</td>\n",
       "      <td>AGGGGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>GGGAGG</td>\n",
       "      <td>uc009vit.3</td>\n",
       "      <td>14362</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>335</td>\n",
       "      <td>18061</td>\n",
       "      <td>24737</td>\n",
       "      <td>uc009vix.2_intron_4_0_chr1_18062_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19516</td>\n",
       "      <td>19528</td>\n",
       "      <td>AAAGCC</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AAGCCA</td>\n",
       "      <td>uc001aai.1</td>\n",
       "      <td>16858</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Exon</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>231</td>\n",
       "      <td>18061</td>\n",
       "      <td>24737</td>\n",
       "      <td>uc009vix.2_intron_4_0_chr1_18062_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>22811</td>\n",
       "      <td>22823</td>\n",
       "      <td>AAAGG</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGAA</td>\n",
       "      <td>uc001aac.4</td>\n",
       "      <td>14362</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td>Non-Promoter</td>\n",
       "      <td>6547</td>\n",
       "      <td>18061</td>\n",
       "      <td>24737</td>\n",
       "      <td>uc009vix.2_intron_4_0_chr1_18062_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-------+\n",
       "| Chromosome   |     Start |       End | class      |    length | Strand       |   num_units | +14   |\n",
       "| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |     (int64) | ...   |\n",
       "|--------------+-----------+-----------+------------+-----------+--------------+-------------+-------|\n",
       "| chr1         |     15240 |     15255 | AGGGCC     |        15 | +            |           2 | ...   |\n",
       "| chr1         |     15383 |     15395 | AGGCGC     |        12 | +            |           2 | ...   |\n",
       "| chr1         |     17476 |     17488 | AGCCG      |        12 | +            |           2 | ...   |\n",
       "| chr1         |     17801 |     17814 | ATCCC      |        13 | +            |           2 | ...   |\n",
       "| chr1         |     18453 |     18466 | AGGCC      |        13 | +            |           2 | ...   |\n",
       "| chr1         |     19412 |     19424 | AGGGGG     |        12 | +            |           2 | ...   |\n",
       "| chr1         |     19516 |     19528 | AAAGCC     |        12 | +            |           2 | ...   |\n",
       "| chr1         |     22811 |     22823 | AAAGG      |        12 | +            |           2 | ...   |\n",
       "+--------------+-----------+-----------+------------+-----------+--------------+-------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 21 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "14 hidden columns: actual_repeat, gene, gene_start, gene_stop, gene_strand, annotation, promoter, dist_to_tss, ... (+ 6 more.)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_intron_pr = str_pr.nearest(intron_pr)\n",
    "nearest_intron_pr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "462ec4ff",
   "metadata": {},
   "source": [
    "Then, get everything "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f49c3333",
   "metadata": {},
   "source": [
    "Next, want to get the eCLIP peaks that are in these STRs. To start, let's try "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc-autonumbering": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
